File: Excel/577 Challenge.py
import pandas as pd
import numpy as np

path = "577 Make ASCII Lamp.xlsx"
test = pd.read_excel(path, header=None, usecols="C:W", skiprows=1, nrows=10)
test = test.fillna("")
test = test.to_numpy()

def centered(matrix, row, how_many):
    pad = (matrix.shape[1] - how_many) // 2
    matrix[row, pad:pad + how_many] = "x"
    return matrix

M = np.full((10, 21), "", dtype=object)

for i in range(1, 10):
    how_many = 1 if i < 3 else 3 if i < 6 else 21 - (i - 6) * 2
    M = centered(M, i, how_many)

print(np.array_equal(M, test)) # True

File: Excel/500-599/577/577 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/577 Make ASCII Lamp.xlsx"
test  = read_excel(path, range = "B2:X11", col_names = F)  %>% as.matrix()
test[is.na(test)] <- ""

centered = function(matrix, row, how_many) {
  pad <- (ncol(matrix) - how_many) %/% 2
  matrix[row, ] <- c(rep("", pad), rep("x", how_many), rep("", ncol(matrix) - how_many - pad))
  matrix
}

M = matrix("", nrow = 10, ncol = 23)

M <- reduce(2:3, ~centered(.x, .y, 1), .init = M)
M <- reduce(4:6, ~centered(.x, .y, 3), .init = M)
M <- reduce(7:10, ~centered(.x, .y, 21 - (.y - 7) * 2), .init = M)

all.equal(M, test, check.attributes = F)
#> [1] TRUE